theano/tensor/nnet/corr_gemm.c

```diff
@@ -0,0 +1,376 @@
+                (Dtype)1., top_diff + top[i]->offset(n) + top_offset * g,
+                col_data + col_offset * g, (Dtype)1.,
+                weight_diff + weight_offset * g);
+            == (see https://github.com/BVLC/caffe/blob/master/src/caffe/util/math_functions.cu#L16)
+            cublasSgemm(CUBLAS_OP_T, CUBLAS_OP_N, K_, M_, N_,
+                1.0,
+                col_data + col_offset * g, N_,
+                top_diff + top[i]->offset(n) + top_offset * g, N_,
+                1.0,
+                weight_diff + weight_offset * g, K_);
+          }
+        }
+        */
+    }
+    else if (direction == 2) {  // backprop wrt. inputs
+        output = bottom;
+        PyArray_FILLWBYTE(bottom, 0);
+        // full convolution: gemm, then col2im
+        // Iterate over batch
+        for (int n = 0; n < batchSize; n++) {
+            // gemm into columns
+            %(gemm)s(&NTrans, &Trans,
+                   &N_, &K_, &M_,
+                   &one,
+                   (%(float_type)s*)PyArray_DATA(top) + n * top_stride, &N_,
+                   (%(float_type)s*)PyArray_DATA(weight), &K_,
+                   &zero,
+                   (%(float_type)s*)PyArray_DATA(col), &N_);
+            // col2im back to the data
+            col2im((%(float_type)s*)PyArray_DATA(col), nChannels, bottomHeight, bottomWidth,
+                    kH, kW, padH, padW, dH, dW, (%(float_type)s*)PyArray_DATA(bottom) + n * bottom_stride);
+        }
+        /*
+        // Original caffe code for comparison
+        // Note that this code was translated from the Theano GPU code,
+        // not the Caffe CPU code.
+        // https://github.com/BVLC/caffe/blob/master/src/caffe/layers/conv_layer.cu
+        for (int n = 0; n < num_; ++n) {
+          // gradient w.r.t. bottom data, if necessary
+          if (propagate_down[i]) {
+            for (int g = 0; g < group_; ++g) {
+              caffe_gpu_gemm<Dtype>(CblasTrans, CblasNoTrans, K_, N_, M_,
+                  (Dtype)1., weight + weight_offset * g,
+                  top_diff + top[i]->offset(n) + top_offset * g,
+                  (Dtype)0., col_diff + col_offset * g);
+              == (see https://github.com/BVLC/caffe/blob/master/src/caffe/util/math_functions.cu#L16)
+              cublasSgemm(CUBLAS_OP_N, CUBLAS_OP_T, N_, K_, M_,
+                  1.,
+                  top_diff + top[i]->offset(n) + top_offset * g, N_,
+                  weight + weight_offset * g, K_,
+                  0.,
+                  col_diff + col_offset * g, N_);
+            }
+            // col2im back to the data
+            col2im_gpu(col_diff, channels_, height_, width_,
+                kernel_h_, kernel_w_, pad_h_, pad_w_, stride_h_, stride_w_,
+                bottom_diff + (*bottom)[i]->offset(n));
+          }
+        }
+        */
+    }
+    // Free temporary columns
+    Py_DECREF(col);
+    if (!bottom_contiguous) {
+	Py_DECREF(bottom);
+    }
```

Comment:
Why did you do it like this? Why you didn't used the numpy function I gave you? I think this work, but it make the code more complicated.

Otherwise the part that make the contig version and decref seem good as the test. I need to leave, was there other think missing? If not, I'll try to do a global last review and run tests locally next week and merge this.

thanks